galley/pkg/config/analysis/analyzers/util/config.go

```diff
@@ -0,0 +1,50 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/meta/metadata"
+	"istio.io/istio/galley/pkg/config/resource"
+)
+
+// MeshConfig returns the mesh configuration object associated with the context
+// Analyzers that call this should include metadata.IstioMeshV1Alpha1MeshConfig as an input in their Metadata
+func MeshConfig(ctx analysis.Context) *v1alpha1.MeshConfig {
+	// Only one MeshConfig should exist in practice, but getting it this way avoids needing
+	// to plumb through the name or enforce/expose a constant.
+	var mc *v1alpha1.MeshConfig
+	ctx.ForEach(metadata.IstioMeshV1Alpha1MeshConfig, func(r *resource.Entry) bool {
+		mc = r.Item.(*v1alpha1.MeshConfig)
+		return true
+	})
+
+	return mc
+}
+
+// IstioNamespace returns the Istio control plane namespace
+// Analyzers that call this should include metadata.IstioMeshV1Alpha1MeshConfig as an input in their Metadata
+func IstioNamespace(ctx analysis.Context) string {
+	// This assumes that rootNamespace is an accurate proxy for the control plane namespace.
```

Comment:
as mentioned in slack, this isn't a great assumption, it's not even the intent of rootNamespace

On a related note, what is the expected behavior when there are multiple control planes in a single cluster? We are hoping to support this pretty well in the future. K guess you would just specify which one to analyze?

Reply:
> as mentioned in slack, this isn't a great assumption, it's not even the intent of rootNamespace

Agreed, but it seems better than hard-coding `istio-system` which is where we're at today. I'm interested in thoughts about what a better approach would look like. I spent a fair bit of time chewing on it, and decided to submit the PR to either make an incremental improvement or simply continue the discussion with more brains.
 
> On a related note, what is the expected behavior when there are multiple control planes in a single cluster? We are hoping to support this pretty well in the future. K guess you would just specify which one to analyze?

Can we assume that each control plane has a namespace, and also somewhere defines a list of data plane namespaces it is responsible for? If so, the expected behavior would be that resources outside the set of control plane + data plane namespaces are ignored. The missing piece currently is having some visible config that maps a control plane namespace to a set of data plane namespaces. e.g. if the mesh config included that list, we could use that.

Reply:
Even hardcoding istio-system is better.. I am pretty sure on the field there is a comment that says `This should NOT be set to the control plane namespace

>Can we assume that each control plane has a namespace
No, they can be in the same namespace

> and also somewhere defines a list of data plane namespaces it is responsible for?
Yes, it will be selected with a label on the namespace (instead of `istio-injection=enabled` it would be something like `istio-environment=reference-to-control-plane`.

> If so, the expected behavior would be that resources outside the set of control plane + data plane namespaces are ignored.

I don't think configs are as bounded by namespace for this to work? Say we have control plane A and B, responsible for namespaces a and b. A DestinationRule in namespace `a` impacts a proxy in namespace `b`. Furthermore a config in the root namespace impacts both.

Reply:
> > Can we assume that each control plane has a namespace
> 
> No, they can be in the same namespace

Hmm, OK. I'd been under the impression that the control plane was addressable by namespace, but apparently that isn't the case. 

> 
> > and also somewhere defines a list of data plane namespaces it is responsible for?
>
> Yes, it will be selected with a label on the namespace (instead of `istio-injection=enabled` it would be something like `istio-environment=reference-to-control-plane`.

If control plane isn't addressable by namespace, what would `reference-to-control-plane` be? It sounds like this is yet to be built, right? Is there a formal definition of "control plane" I should be using going forward? 

> 
> > If so, the expected behavior would be that resources outside the set of control plane + data plane namespaces are ignored.
> 
> I don't think configs are as bounded by namespace for this to work? Say we have control plane A and B, responsible for namespaces a and b. A DestinationRule in namespace `a` impacts a proxy in namespace `b`. Furthermore a config in the root namespace impacts both.

That's true, so instead we'd want to restrict the validation messages to the scoped namespaces while keeping everything in the cluster visible in the snapshot used for analysis. That should be doable, although it may make the `--all-namespaces` option a bit confusing. (i.e. should it be all namespaces in a shared control plane, or all namespaces cluster-wide? In the latter case, we'd need to be very clever about how to make sure we're using the right meshConfig / mesh gateways for a particular data plane resource we're analyzing. Sounds ugly.)

Reply:
Its not really yet to be built, it works today but isn't well documented or tested. Its literally a 1 line change, changing https://github.com/istio/installer/blob/master/istio-control/istio-autoinject/templates/mutatingwebhook.yaml#L18-L19 to point to the control plane and configuring the relevant match rules in that webhook. The injection controls which control plane is used -- it injects the address of pilot and mixer, etc

Reply:
Resolved for this analyzer by looking for the `istio=` label instead of relying on namespace.